packages/core/src/resolver.ts

```diff
@@ -1,195 +1,366 @@
+  kl: PipeFn<K, L, CK, CL>,
+): (input: A, ctx: CL) => L
+function pipe<
+  A,
+  B,
+  C,
+  D,
+  E,
+  F,
+  G,
+  H,
+  I,
+  J,
+  K,
+  L,
+  M,
+  CA,
+  CB = CA,
+  CC = CB,
+  CD = CC,
+  CE = CD,
+  CF = CE,
+  CG = CF,
+  CH = CG,
+  CI = CH,
+  CJ = CI,
+  CK = CJ,
+  CL = CK,
+  CM = CL
+>(
+  ab: PipeFn<A, B, CA, CB>,
+  bc: PipeFn<B, C, CB, CC>,
+  cd: PipeFn<C, D, CC, CD>,
+  de: PipeFn<D, E, CD, CE>,
+  ef: PipeFn<E, F, CE, CF>,
+  fg: PipeFn<F, G, CF, CG>,
+  gh: PipeFn<G, H, CG, CH>,
+  hi: PipeFn<H, I, CH, CI>,
+  ij: PipeFn<I, J, CI, CJ>,
+  jk: PipeFn<J, K, CJ, CK>,
+  kl: PipeFn<K, L, CK, CL>,
+  lm: PipeFn<L, M, CL, CM>,
+): (input: A, ctx: CL) => M
+// function pipe<A, B, C, D, E, F, G, H, I, J, K, L, M, N>(
+//   ab: PipeFn<A, B>,
+//   bc: PipeFn<B, C>,
+//   cd: PipeFn<C, D>,
+//   de: PipeFn<D, E>,
+//   ef: PipeFn<E, F>,
+//   fg: PipeFn<F, G>,
+//   gh: PipeFn<G, H>,
+//   hi: PipeFn<H, I>,
+//   ij: PipeFn<I, J>,
+//   jk: PipeFn<J, K>,
+//   kl: PipeFn<K, L>,
+//   lm: PipeFn<L, M>,
+//   mn: PipeFn<M, N>,
+// ): (input: A, ctx: Ctx) => N
+// function pipe<A, B, C, D, E, F, G, H, I, J, K, L, M, N, O>(
+//   ab: PipeFn<A, B>,
+//   bc: PipeFn<B, C>,
+//   cd: PipeFn<C, D>,
+//   de: PipeFn<D, E>,
+//   ef: PipeFn<E, F>,
+//   fg: PipeFn<F, G>,
+//   gh: PipeFn<G, H>,
+//   hi: PipeFn<H, I>,
+//   ij: PipeFn<I, J>,
+//   jk: PipeFn<J, K>,
+//   kl: PipeFn<K, L>,
+//   lm: PipeFn<L, M>,
+//   mn: PipeFn<M, N>,
+//   no: PipeFn<N, O>,
+// ): (input: A, ctx: Ctx) => O
+// function pipe<A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P>(
+//   ab: PipeFn<A, B>,
+//   bc: PipeFn<B, C>,
+//   cd: PipeFn<C, D>,
+//   de: PipeFn<D, E>,
+//   ef: PipeFn<E, F>,
+//   fg: PipeFn<F, G>,
+//   gh: PipeFn<G, H>,
+//   hi: PipeFn<H, I>,
+//   ij: PipeFn<I, J>,
+//   jk: PipeFn<J, K>,
+//   kl: PipeFn<K, L>,
+//   lm: PipeFn<L, M>,
+//   mn: PipeFn<M, N>,
+//   no: PipeFn<N, O>,
+//   op: PipeFn<O, P>,
+// ): (input: A, ctx: Ctx) => P
+// function pipe<A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q>(
+//   ab: PipeFn<A, B>,
+//   bc: PipeFn<B, C>,
+//   cd: PipeFn<C, D>,
+//   de: PipeFn<D, E>,
+//   ef: PipeFn<E, F>,
+//   fg: PipeFn<F, G>,
+//   gh: PipeFn<G, H>,
+//   hi: PipeFn<H, I>,
+//   ij: PipeFn<I, J>,
+//   jk: PipeFn<J, K>,
+//   kl: PipeFn<K, L>,
+//   lm: PipeFn<L, M>,
+//   mn: PipeFn<M, N>,
+//   no: PipeFn<N, O>,
+//   op: PipeFn<O, P>,
+//   pq: PipeFn<P, Q>,
+// ): (input: A, ctx: Ctx) => Q
 function pipe(...args: unknown[]): unknown {
-  const functions = args as PipeFn<unknown, unknown>[]
+  const functions = args as PipeFn<unknown, unknown, Ctx>[]
 
   return async function (input: unknown, ctx: Ctx) {
     let lastResult = input
     for (let fn of functions) {
       lastResult = await fn(lastResult, ctx)
+      if (isResultWithContext(lastResult)) {
+        ctx = lastResult.ctx as Ctx
+        lastResult = lastResult.value
+      }
     }
     return lastResult
   }
 }
 
+interface ResolverAuthorize {
```

Comment:
I declared the type of `resolver.authorize`, because it was getting pretty complex, and while "types first, implementation later, functions as interfaces" is probably the most tiring way to write code, it results in the shortest error messages and tooltips

![image](https://user-images.githubusercontent.com/15332326/106330899-5bc87100-6284-11eb-84af-1c4f011b7c47.png)

Reply:
@hasparus Very interesting comment. 

Where can I read more about the "types first, implementation later, functions as interfaces" philosophy?

I understand "types first, implementation later". But what exactly does "functions as interfaces" mean in this context? How does this fit together with the fact that `PipeFn` is also a function, but you defined it as type alias and not as `interface`?